packages/at_secondary_server/lib/src/verb/handler/update_meta_verb_handler.dart

```diff
@@ -24,14 +25,45 @@ class UpdateMetaVerbHandler extends AbstractUpdateVerbHandler {
       Response response,
       HashMap<String, String?> verbParams,
       InboundConnection atConnection) async {
-    var updatePreProcessResult =
-        await super.preProcessAndNotify(response, verbParams, atConnection);
-
-    // update the key in data store
-    logger.finer(
-        'calling keyStore.putMeta(${updatePreProcessResult.atKey}, ${updatePreProcessResult.atData.metaData!}');
-    var result = await keyStore.putMeta(
-        updatePreProcessResult.atKey, updatePreProcessResult.atData.metaData!);
-    response.data = result?.toString();
+    UpdateParams updateParams = getUpdateParams(verbParams);
+
+    String dataStoreKey = getDataStoreKey(updateParams);
+
+    updateMutexes.putIfAbsent(dataStoreKey, () => (Mutex(), 0));
+
+    try {
+      var mutexRecord = updateMutexes[dataStoreKey]!;
+      updateMutexes[dataStoreKey] = (mutexRecord.$1, mutexRecord.$2 + 1);
+      logger.finest(
+          'Acquiring mutex for $dataStoreKey - ${mutexRecord.$2 + 1} waiting (including me)');
+      await mutexRecord.$1.acquire();
+
+      var updatePreProcessResult = await super.preProcessAndNotify(
+        response,
+        verbParams,
+        updateParams,
+        atConnection,
+      );
+
+      // update the key in data store
+      logger.finer(
+          'calling keyStore.putMeta(${updatePreProcessResult.atKey}, ${updatePreProcessResult.atData.metaData!}');
+      var result = await keyStore.putMeta(updatePreProcessResult.atKey,
+          updatePreProcessResult.atData.metaData!);
+      response.data = result?.toString();
+    } finally {
+      String logMsg = 'Releasing mutex on $dataStoreKey';
+      var mutexRecord = updateMutexes[dataStoreKey]!;
+      mutexRecord.$1.release();
+      updateMutexes[dataStoreKey] = (mutexRecord.$1, mutexRecord.$2 - 1);
+      if (updateMutexes[dataStoreKey]!.$2 == 0) {
+        logger.finest(
+            '$logMsg : 0 now waiting for mutex on $dataStoreKey - removing mutex');
+        updateMutexes.remove(dataStoreKey);
+      } else {
+        logger.finest(
+            '$logMsg : ${updateMutexes[dataStoreKey]!.$2} still waiting for mutex on $dataStoreKey');
+      }
+    }
```

Comment:
as mentioned above, the `finally` block here
- releases the mutex
- updates the waiting count in the `updateMutexes` map entry
- if the waiting count is 0 then remove the entry from the map